src/explorer.ts

```diff
@@ -270,24 +275,142 @@ export class OpenShiftExplorer implements TreeDataProvider<ExplorerItem>, Dispos
+        if (notRunning.length === 0) {
+            return '-green'; // All running - return 'green'
+        }
+        // Find any 'Failed' or 'Unknown' pod - if any return error ('red')
+        const failed = notRunning.find((pod) => pod.status &&
+                (pod.status.phase === 'Failed' || pod.status.phase === 'Unknown'));
+        if (failed) {
+            return '-red'; // At least one failed or unknown - return 'red'
+        }
+        // Find any 'Pending' pod - if any return pending ('yellow')
+        const pending = notRunning.find((pod) => pod.status && pod.status.phase === 'Pending');
+        return pending ? '-yellow' : '';
+    }
+
+    /*
+     * Search for 'CrashLoopBackOff` in Pod's opbect:
+     * status:
+     *   containerStatuses:
+     *     state:
+     *       waiting:
+     *         message: |
+     *           container create failed: time="2024-05-28T10:30:28Z" level=error msg="runc create failed: unable to start container process: exec: \"asdf\": executable file not found in $PATH"
+     *         reason: CreateContainerError
+     */
+    private detectCrashLoopBackOff(pods): DeploymentPodErrors {
+        let inCrashLoopBackOff = false;
+        const messages: string[] = [];
+
+        // Search for Pod continers' errors
+        pods.forEach((pod) => {
+            pod.status?.containerStatuses &&
+            pod.status.containerStatuses.forEach((cs) => {
+                if (cs.state?.waiting) {
+                    const reason = cs.state.waiting.reason;
+                    const message = cs.state.waiting.message;
+
+                    inCrashLoopBackOff = inCrashLoopBackOff || reason === 'CrashLoopBackOff';
+
+                    const msg = `${reason}: ${message ? message.trim(): 'No any valuable message'}`;
+                    // Skip duplicates and show not more than 10 errors
+                    if (messages.length <= 10 && !(messages.find((m) => m === msg))) {
+                        messages.push(msg);
+                    }
+                }
+            });
+        });
+
+        return {
+            inCrashLoopBackOff,
+            messages
+        }
+    }
+
+    private collectDeploymentErrors(deployment): string[] {
+        const messages: string[] = [];
+        deployment.status.conditions.filter((c) => c.status === 'False')
+            .forEach((c) => {
+                const message = `${c.reason}: ${c.message ? c.message.trim(): 'No any valuable message'}`;
+
+                // Skip duplicates and show not more than 10 errors
+                if (messages.length <= 10 && !(messages.find((m) => m === message))) {
+                    messages.push(message);
+                }
+            });
+
+        return messages;
+    }
+
+    private async getDeploymentItem(element): Promise<TreeItem> {
+        const shouldHaveReplicas = element.spec.replicas > 0;
+        const availableReplicas = element.status.availableReplicas ? element.status.availableReplicas :
```

Comment:
From my understanding the two numbers here would best be labelled "desiredReplicas" and "actualReplicas". "desiredReplicas" should be take from the `spec` section, and "actualReplicas" should be taken from the `status` section. Does this make sense to you? If not I might misunderstand something.

I think the difference between `status.availableReplicas` vs `status.readyReplicas` vs `status.replicas` is one means it's in the process of being created, one means it's has been created and is running, and one means it's been created and running and has a [probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/) configured that says it's healthy.

Reply:
So, OK. I take `desiredReplicas = element.spec.replicas` and don't try searching for `element.status.availableReplicas`

For the `actualReplicas` I take `element.status.replicas` value or 0 if it doesn't exist (which is possible, f.i., when scaled to 0).

But `element.status.replicas` tells nothing on real state of the replicas, as for a failing `pet-alpine` deployment (on Sandbox) we have:
```
status: {
  ...
  replicas: 1, 
  unavailableReplicas: 1, 
  updatedReplicas: 1
  ...
}
```

So, we actually have ONE replica and this the only replica is UNAVAILABLE (not working)...

... but we'll still show a tooltip like:

```
deployment: my-pet-alpine
desiredReplicas: 1
actualReplicas: 1
...
```
So user will think that everything should be OK with this replica - it's started and should be `Running`, while it's actually not (here it's `Pending`).

Probably we should show more replicas-related counters then, shouldn't we?

Reply:
I like this behaviour!